File: backend/app/playlist_imports.py
"""The playlist-import pipelines: fetch and match, download, and retry.

Mirrors app/jobs.py — plain sync functions, run by a Celery worker, that own
their session and record every failure on the row instead of raising. Spotify
is only contacted during FETCHING; from MATCHING onward an expired token or
lapsed Premium can no longer affect the run (ADR-005).
"""

import logging
import time
from dataclasses import asdict

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.config import get_settings
from app.db import SessionLocal
from app.events import playlist_import_events
from app.installs import owned_by
from app.jobs import run_import_job
from app.matching import CANDIDATE_LIMIT, build_search_query, classify, score_candidates
from app.models import (
    ImportJob,
    ImportStatus,
    Playlist,
    PlaylistImport,
    PlaylistImportStatus,
    PlaylistItem,
    Song,
    SpotifyAccount,
    TrackMatch,
    TrackMatchStatus,
)
from app.pacing import AdaptivePacer, retry_backoff_seconds
from app.schemas import PlaylistImportRead
from app.spotify import SpotifyError, fetch_all_playlist_tracks, fetch_all_saved_tracks
from app.spotify_accounts import get_fresh_access_token
from app.ytdlp import (
    ExtractionError,
    SearchResult,
    TransientExtractionError,
    display_artist,
    list_playlist,
    search,
)

logger = logging.getLogger(__name__)

# external_playlist_id sentinel for the user's Liked Songs.
LIKED_SONGS_ID = "liked"

# Match statuses the confirm phase will download.
IMPORTABLE_STATUSES = {TrackMatchStatus.AUTO_MATCHED, TrackMatchStatus.ACCEPTED}

# Import statuses that mean "the server was mid-run" — as opposed to `review`,
# which waits on a person and survives restarts untouched.
INTERRUPTIBLE_STATUSES = (
    PlaylistImportStatus.QUEUED,
    PlaylistImportStatus.FETCHING,
    PlaylistImportStatus.MATCHING,
    PlaylistImportStatus.IMPORTING,
)


def publish_import(imp: PlaylistImport) -> None:
    """Tell WebSocket listeners the run moved.

    Public since #353: the candidates endpoint advances an import from a request
    handler rather than from the pipeline, and a review screen watching the
    socket should not have to poll to notice.
    """
    playlist_import_events.publish(
        imp.id, PlaylistImportRead.model_validate(imp).model_dump(mode="json")
    )


def _advance(imp: PlaylistImport, status: PlaylistImportStatus, db: Session) -> None:
    """Move the import to `status`, persist it, then tell WebSocket listeners.
    Commit first — the DB is the source of truth (ADR-004)."""
    imp.status = status
    db.commit()
    db.refresh(imp)
    logger.info(
        "playlist import status",
        extra={
            "playlist_import_id": imp.id,
            "status": status.value,
            "imported": imp.imported_count,
            "failed": imp.failed_count,
        },
    )
    publish_import(imp)


def run_playlist_import(playlist_import_id: int) -> None:
    """Drive an import to `review` (or `failed`). Dispatches on the source:
    Spotify fetches then matches; YouTube just fetches (its entries are already
    videos, so there's nothing to match) — see ADR-010.

    One run does **not** reach review here: a Spotify import flagged
    `client_matches` stops after fetching, because the searching happens on the
    phone (#353). `POST /playlist-imports/{id}/candidates` finishes it."""
    db = SessionLocal()
    try:
        imp = db.get(PlaylistImport, playlist_import_id)
        if imp is None:
            return

        try:
            if imp.service == "youtube":
                _fetch_youtube_playlist(imp, db)
            else:
                _fetch_and_match_spotify(imp, db)
            # A client-matching run is not finished when this returns — it has
            # fetched the tracklist and is waiting for the phone to post what it
            # found (#353). `POST /{id}/candidates` is what moves it to review,
            # once every track has been answered for.
            if not imp.client_matches:
                _advance(imp, PlaylistImportStatus.REVIEW, db)
        except Exception as exc:
            logger.exception("playlist import failed", extra={"error": str(exc)})
            imp.error = str(exc)
            _advance(imp, PlaylistImportStatus.FAILED, db)
    finally:
        db.close()


def _fetch_and_match_spotify(imp: PlaylistImport, db: Session) -> None:
    settings = get_settings()
    account = db.get(SpotifyAccount, imp.account_id) if imp.account_id else None
    if account is None:
        raise SpotifyError("The Spotify account for this import is no longer connected")

    _advance(imp, PlaylistImportStatus.FETCHING, db)
    token = get_fresh_access_token(db, account, settings.spotify_client_id or "")
    if imp.external_playlist_id == LIKED_SONGS_ID:
        tracks = fetch_all_saved_tracks(token)
    else:
        tracks = fetch_all_playlist_tracks(token, imp.external_playlist_id)

    for position, track in enumerate(tracks):
        db.add(
            TrackMatch(
                playlist_import_id=imp.id,
                position=position,
                external_id=track.external_id,
                title=track.title,
                artist=track.artist,
                album=track.album,
                duration_s=track.duration_s,
            )
        )
    imp.track_count = len(tracks)
    logger.info("fetched tracks from spotify", extra={"track_count": len(tracks)})
    # Spotify's part ends here — nothing below needs the token.
    _advance(imp, PlaylistImportStatus.MATCHING, db)

    # The client is doing the searching, so the run stops here (#353).
    #
    # It stays in MATCHING rather than gaining a status of its own: from the
    # outside "matching" is exactly what is happening, and the only difference
    # is which machine is doing it. `POST /{id}/candidates` drives the rest, and
    # the startup sweep already treats a stalled MATCHING run as interrupted.
    if imp.client_matches:
        logger.info("waiting for client-supplied candidates", extra={"import_id": imp.id})
        return

    matches = db.scalars(
        select(TrackMatch)
        .where(TrackMatch.playlist_import_id == imp.id)
        .order_by(TrackMatch.position)
    ).all()
    for match in matches:
        try:
            results = search(build_search_query(match.title, match.artist), CANDIDATE_LIMIT)
        except ExtractionError as exc:
            # One unsearchable track degrades that row, never the run.
            logger.warning(
                "track search failed",
                extra={"track": match.title, "artist": match.artist, "error": str(exc)},
            )
            match.status = TrackMatchStatus.NO_MATCH
            match.error = str(exc)
        else:
            record_candidates(match, results)
        imp.matched_count += 1
        db.commit()
        db.refresh(imp)
        publish_import(imp)


def record_candidates(match: TrackMatch, results: list[SearchResult]) -> None:
    """Score `results` against `match` and write the outcome onto the row.

    Split out of the matching loop by #353 so that candidates found **on a
    phone** are scored by exactly the same code as candidates the server found
    itself. That is the whole point of the design: the search moves, because
    only a residential address is allowed to do it, and the scorer does not,
    because there is no reason for it to and two scorers would drift.

    Pure apart from the assignment — no session, no network — so it is callable
    from the pipeline and from a request handler alike.
    """
    scored = score_candidates(match.title, match.artist, match.duration_s, results)
    match.candidates = [asdict(candidate) for candidate in scored]
    if scored:
        match.chosen_url = scored[0].url
        match.confidence = scored[0].score
    # Cleared on purpose: a row being re-matched after a failure kept its old
    # message, so a track that succeeded on the second attempt still showed the
    # reason the first one failed.
    match.error = None
    match.status = classify(scored[0].score if scored else None)


def _fetch_youtube_playlist(imp: PlaylistImport, db: Session) -> None:
    """List a public YouTube playlist and record each video as an already-matched
    track — no search stage. The confirm/download chain then runs unchanged."""
    _advance(imp, PlaylistImportStatus.FETCHING, db)
    listing = list_playlist(imp.external_playlist_id)
    if listing.title:
        imp.name = listing.title

    for position, entry in enumerate(listing.entries):
        db.add(
            TrackMatch(
                playlist_import_id=imp.id,
                position=position,
                external_id=None,
                title=entry.title,
                # Not the raw channel: an auto-generated Topic channel is
                # "<Artist> - Topic", and some of them name nobody at all (#307).
                artist=display_artist(entry.uploader) or "Unknown",
                album=None,
                duration_s=entry.duration,
                # The video is its own (single) candidate; no machine confidence.
                candidates=[
                    {
                        "url": entry.url,
                        "title": entry.title,
                        "uploader": entry.uploader,
                        "duration": entry.duration,
                        "thumbnail": entry.thumbnail,
                        "score": None,
                    }
                ],
                chosen_url=entry.url,
                confidence=None,
                status=TrackMatchStatus.AUTO_MATCHED,
            )
        )
    # ⚠️ The playlist's size, not the number of entries that survived (#585).
    #
    # `list_playlist` drops an entry with no title — which is what a deleted,
    # private or region-blocked video looks like to flat extraction — and this
    # line used to record the survivors as the import's own total. An
    # eighteen-track playlist then *was* a four-track import as far as anything
    # downstream could tell, and the phone honestly reported "4/4 done" about a
    # number that had already been redefined.
    #
    # `matched_count` stays the entries we actually hold, so the pair now says
    # "4 of 18" rather than "4 of 4".
    imp.track_count = listing.announced_count
    imp.matched_count = len(listing.entries)
    if listing.skipped:
        # Warning, not info: this is the difference between the playlist the
        # user is looking at and the one they are about to import, and it is the
        # only place that difference is observable.
        #
        # Not raised. ADR-013 refuses a short read for NetEase, QQ and Kugou,
        # and rightly — there a shortfall means the *request* half worked. Here
        # it usually means the videos are genuinely gone, and refusing would
        # make a playlist with one deleted track unimportable. So: report.
        logger.warning(
            "youtube playlist entries skipped",
            extra={
                "announced": listing.announced_count,
                "usable": len(listing.entries),
                "skipped": listing.skipped,
            },
        )
    logger.info("fetched youtube playlist", extra={"track_count": len(listing.entries)})


def _recount(imp: PlaylistImport, db: Session) -> None:
    """Recompute the progress counters from the rows themselves.

    Derived rather than incremented, so a retry that turns a failure into a
    success can't leave the totals lying.
    """
    # Sessions here are autoflush=False, so the status just set on the row in
    # memory has to reach the database before we count rows in it.
    db.flush()
    counts = dict(
        db.execute(
            select(TrackMatch.status, func.count())
            .where(TrackMatch.playlist_import_id == imp.id)
            .group_by(TrackMatch.status)
        ).all()
    )
    imp.imported_count = counts.get(TrackMatchStatus.IMPORTED, 0)
    imp.failed_count = counts.get(TrackMatchStatus.FAILED, 0)


def _next_playlist_position(playlist_id: int, db: Session) -> int:
    highest = db.scalar(
        select(func.max(PlaylistItem.position)).where(PlaylistItem.playlist_id == playlist_id)
    )
    return 0 if highest is None else highest + 1


def _import_match(
    match: TrackMatch, imp: PlaylistImport, db: Session, pacer: AdaptivePacer
) -> None:
    """Download one accepted match and, on success, append it to the playlist.

    Shared by the first confirmed run and by retries, so both behave
    identically. Records failures on the row and returns — a bad track never
    stops the batch. The `pacer` carries the adaptive inter-download delay
    across calls in a batch (B0).
    """
    song_id = None
    if match.chosen_url is None:
        match.status = TrackMatchStatus.FAILED
        match.error = "No chosen URL"
    else:
        # Reuse a song that's already in the library (same source) instead of
        # downloading the same audio twice.
        # Only reuse a song this same owner already has: matching on URL alone
        # would hand one person's file to another's playlist (#170).
        existing = db.scalar(
            select(Song).where(
                Song.source_url == match.chosen_url,
                owned_by(Song, imp.owner_install_id),
            )
        )
        if existing is not None:
            song_id = existing.id
        else:
            job = ImportJob(
                source_url=match.chosen_url,
                status=ImportStatus.QUEUED,
                owner_install_id=imp.owner_install_id,
            )
            db.add(job)
            # Commit ends our transaction before the inline call: run_import_job
            # commits on its own session, and two sessions must never hold
            # SQLite write locks at once.
            db.commit()
            # Single jobs get their retries from the Celery task layer, but a
            # batch calls the pipeline inline — so retry transient failures here
            # too, with the same bounded backoff, instead of failing the track
            # on the first throttle (B3).
            max_retries = get_settings().download_max_retries
            hit_transient = False
            for attempt in range(max_retries + 1):
                try:
                    run_import_job(job.id)
                    break  # returned (DONE, or a permanent failure it recorded)
                except TransientExtractionError as exc:
                    hit_transient = True
                    if attempt >= max_retries:
                        logger.warning(
                            "giving up on track after retries",
                            extra={
                                "track": match.title,
                                "attempts": attempt + 1,
                                "error": str(exc),
                            },
                        )
                        break
                    delay = retry_backoff_seconds(attempt)
                    logger.info(
                        "retrying track after transient failure",
                        extra={
                            "track": match.title,
                            "attempt": attempt + 1,
                            "retry_in_seconds": delay,
                        },
                    )
                    time.sleep(delay)

            db.expire_all()
            db.refresh(job)
            match.import_job_id = job.id
            if job.status == ImportStatus.DONE and job.song_id is not None:
                song_id = job.song_id
            else:
                match.status = TrackMatchStatus.FAILED
                match.error = job.error or "Download failed"
            # Pacer (B0): a clean download speeds us up; any throttling we saw —
            # even if we recovered from it — says back off; a private/removed
            # video is neutral, since it says nothing about the rate.
            if hit_transient:
                pacer.on_transient_failure()
            elif job.status == ImportStatus.DONE:
                pacer.on_success()
            pacer.wait()

    if song_id is not None:
        logger.info(
            "track imported",
            extra={"track": match.title, "song_id": song_id, "position": match.position},
        )
        match.song_id = song_id
        match.status = TrackMatchStatus.IMPORTED
        match.error = None
        db.add(
            PlaylistItem(
                playlist_id=imp.playlist_id,
                song_id=song_id,
                position=_next_playlist_position(imp.playlist_id, db),
            )
        )

    _recount(imp, db)
    db.commit()
    db.refresh(imp)
    publish_import(imp)


def run_confirmed_import(playlist_import_id: int) -> None:
    """Phase two: download every accepted match through the Phase-1 pipeline
    and mirror the successes into a new local playlist.

    Sequential on purpose — polite to YouTube, and the rate limit on the job
    task paces it further. Per-track failures are recorded on the row and
    skipped; the run always ends `done` unless the orchestration itself blows
    up (partial success is success, and visible).
    """
    db = SessionLocal()
    try:
        imp = db.get(PlaylistImport, playlist_import_id)
        if imp is None:
            return

        try:
            playlist = Playlist(name=imp.name, owner_install_id=imp.owner_install_id)
            db.add(playlist)
            db.flush()
            imp.playlist_id = playlist.id
            db.commit()
            db.refresh(imp)
            publish_import(imp)

            matches = db.scalars(
                select(TrackMatch)
                .where(
                    TrackMatch.playlist_import_id == imp.id,
                    TrackMatch.status.in_(IMPORTABLE_STATUSES),
                )
                .order_by(TrackMatch.position)
            ).all()

            pacer = AdaptivePacer.from_settings()
            for match in matches:
                _import_match(match, imp, db, pacer)

            _advance(imp, PlaylistImportStatus.DONE, db)
        except Exception as exc:
            logger.exception("playlist import failed", extra={"error": str(exc)})
            imp.error = str(exc)
            _advance(imp, PlaylistImportStatus.FAILED, db)
    finally:
        db.close()


def run_retry_failed_matches(playlist_import_id: int, match_ids: list[int]) -> None:
    """Have another go at tracks that failed to download.

    Most download failures are transient (throttling), so this exists to make
    them recoverable without re-importing the whole playlist. Recovered songs
    are appended to the playlist the import already created.
    """
    db = SessionLocal()
    try:
        imp = db.get(PlaylistImport, playlist_import_id)
        if imp is None or imp.playlist_id is None:
            return

        try:
            matches = db.scalars(
                select(TrackMatch)
                .where(
                    TrackMatch.playlist_import_id == imp.id,
                    TrackMatch.id.in_(match_ids),
                    TrackMatch.status == TrackMatchStatus.FAILED,
                )
                .order_by(TrackMatch.position)
            ).all()
            if not matches:
                _advance(imp, PlaylistImportStatus.DONE, db)
                return

            # Back to accepted while they're in flight, so the counters (and
            # the progress bar) describe what's actually happening.
            for match in matches:
                match.status = TrackMatchStatus.ACCEPTED
                match.error = None
            _recount(imp, db)
            db.commit()
            db.refresh(imp)
            publish_import(imp)

            pacer = AdaptivePacer.from_settings()
            for match in matches:
                _import_match(match, imp, db, pacer)

            _advance(imp, PlaylistImportStatus.DONE, db)
        except Exception as exc:
            logger.exception("playlist import failed", extra={"error": str(exc)})
            imp.error = str(exc)
            _advance(imp, PlaylistImportStatus.FAILED, db)
    finally:
        db.close()


def fail_interrupted_imports() -> None:
    """Mark imports that were mid-run when the server last stopped as failed.

    Celery redelivers *queued* work after a crash, but whatever a worker was
    running when it died is gone. Without this sweep (called from the lifespan
    handler) such an import would sit in a non-terminal state forever, looking
    alive.
    """
    db = SessionLocal()
    try:
        stuck = db.scalars(
            select(PlaylistImport).where(PlaylistImport.status.in_(INTERRUPTIBLE_STATUSES))
        ).all()
        for imp in stuck:
            imp.error = "Interrupted by a server restart"
            imp.status = PlaylistImportStatus.FAILED
        if stuck:
            db.commit()
    finally:
        db.close()

File: mobile/src/library/playlistImports.ts
/**
 * Playlist imports and their track matches, on the device (#610).
 *
 * The storage half of moving review-based imports off the server (#608). The
 * *hooks* that read this are #611; this file is only the rows.
 *
 * Mirrors `backend/app/models.py`'s `PlaylistImport` and `TrackMatch` closely
 * enough that `MatchReview.tsx` does not have to learn a new shape — the review
 * screen is 632 lines and none of it is about where the rows live.
 *
 * ## Three server columns are deliberately absent
 *
 * `account_id`, `client_matches` and `owner_install_id`. See the v10 migration
 * in `db.ts` for why each one stops meaning anything on a device.
 *
 * ⚠️ **Statuses are the lower-case *wire* spellings** — `'needs_review'`, not
 * the `NEEDS_REVIEW` the backend's own table holds. That is what mobile has
 * always received and what the review screen compares against.
 */

import { randomHex } from '../random'
import { openLibraryDb, withLibraryTransaction } from './db'
import { createPlaylist } from './playlists'

const ID_BYTES = 16

/** Where a review-based import came from. */
export type ImportService = 'spotify' | 'netease' | 'qq' | 'kugou' | 'youtube'

export type LocalImportStatus =
  'queued' | 'fetching' | 'matching' | 'review' | 'importing' | 'done' | 'failed'

export type LocalTrackMatchStatus =
  | 'pending'
  | 'auto_matched'
  | 'needs_review'
  | 'no_match'
  | 'accepted'
  | 'rejected'
  | 'imported'
  | 'failed'

/** One scored alternative for a track. Stored as JSON on the row. */
export interface LocalMatchCandidate {
  url: string
  title: string
  uploader: string | null
  duration: number | null
  thumbnail?: string | null
  /** Machine score 0..1, or null for a candidate that was never scored — a
   *  YouTube-playlist entry is its own candidate (ADR-010). */
  score: number | null
  /** Which platform produced it (#551), so a mixed list is readable. */
  source?: string
}

export interface LocalPlaylistImport {
  id: string
  service: ImportService
  external_playlist_id: string
  name: string
  status: LocalImportStatus
  track_count: number | null
  matched_count: number
  import_total: number | null
  imported_count: number
  failed_count: number
  error: string | null
  playlist_id: string | null
  created_at: string
  updated_at: string
}

export interface LocalTrackMatch {
  id: string
  playlist_import_id: string
  position: number
  external_id: string | null
  title: string
  artist: string
  album: string | null
  duration_s: number | null
  candidates: LocalMatchCandidate[]
  chosen_url: string | null
  confidence: number | null
  status: LocalTrackMatchStatus
  song_id: string | null
  error: string | null
}

/** One track as the fetcher hands it over, before anything has been searched. */
export interface TrackToMatch {
  external_id?: string | null
  title: string
  artist: string
  album?: string | null
  duration_s?: number | null
}

/** The stored row, before `candidates` is parsed back out of its JSON. */
type TrackMatchRow = Omit<LocalTrackMatch, 'candidates'> & { candidates: string }

function now(): string {
  return new Date().toISOString()
}

/**
 * Parse a stored `candidates` blob.
 *
 * ⚠️ **Tolerant on purpose.** A row whose JSON cannot be read is a row whose
 * *alternatives* are lost, which is a degraded review screen; throwing here
 * would instead make the whole import unopenable and lose the tracks as well.
 * The server's column had a default of `[]` for the same reason.
 */
function parseCandidates(raw: string): LocalMatchCandidate[] {
  try {
    const parsed: unknown = JSON.parse(raw)
    return Array.isArray(parsed) ? (parsed as LocalMatchCandidate[]) : []
  } catch {
    return []
  }
}

function hydrate(row: TrackMatchRow): LocalTrackMatch {
  return { ...row, candidates: parseCandidates(row.candidates) }
}

export async function createImport(input: {
  service: ImportService
  externalPlaylistId: string
  name: string
  trackCount?: number | null
}): Promise<string> {
  const db = await openLibraryDb()
  const id = randomHex(ID_BYTES)
  const timestamp = now()
  await db.runAsync(
    `INSERT INTO playlist_imports
       (id, service, external_playlist_id, name, status, track_count, created_at, updated_at)
     VALUES (?, ?, ?, ?, 'queued', ?, ?, ?)`,
    [
      id,
      input.service,
      input.externalPlaylistId,
      input.name,
      input.trackCount ?? null,
      timestamp,
      timestamp,
    ],
  )
  return id
}

export async function getImport(importId: string): Promise<LocalPlaylistImport | null> {
  const db = await openLibraryDb()
  return db.getFirstAsync<LocalPlaylistImport>(`SELECT * FROM playlist_imports WHERE id = ?`, [
    importId,
  ])
}

/** Newest first, which is the order the imports list has always shown. */
export async function listImports(): Promise<LocalPlaylistImport[]> {
  const db = await openLibraryDb()
  return db.getAllAsync<LocalPlaylistImport>(
    `SELECT * FROM playlist_imports ORDER BY created_at DESC`,
  )
}

/**
 * Patch an import row. Every caller sets `updated_at`, so it is set here rather
 * than by each of them — the server's `onupdate` did the same job.
 */
export async function updateImport(
  importId: string,
  patch: Partial<
    Pick<
      LocalPlaylistImport,
      | 'status'
      | 'name'
      | 'track_count'
      | 'matched_count'
      | 'import_total'
      | 'imported_count'
      | 'failed_count'
      | 'error'
      | 'playlist_id'
    >
  >,
): Promise<void> {
  const columns = Object.keys(patch) as (keyof typeof patch)[]
  if (columns.length === 0) return
  const db = await openLibraryDb()
  const assignments = columns.map((column) => `${column} = ?`).join(', ')
  await db.runAsync(`UPDATE playlist_imports SET ${assignments}, updated_at = ? WHERE id = ?`, [
    ...columns.map((column) => patch[column] ?? null),
    now(),
    importId,
  ])
}

/**
 * Write the fetched tracklist as `pending` matches.
 *
 * ⚠️ **`track_count` is set from what was actually stored**, never from what
 * the fetcher claimed. #585 was exactly this: `list_playlist` dropped entries it
 * could not read and recorded the survivors as the playlist's size, so an
 * eighteen-track import reported "4/4 done". The count and the playlist are
 * different facts, and only one of them is a row you can point at.
 *
 * One transaction: a half-written tracklist would resume as a silently short
 * import.
 */
export async function addTracks(
  importId: string,
  tracks: readonly TrackToMatch[],
): Promise<number> {
  const timestamp = now()
  const db = await openLibraryDb()
  await withLibraryTransaction(db, async () => {
    for (const [position, track] of tracks.entries()) {
      await db.runAsync(
        `INSERT INTO track_matches
           (id, playlist_import_id, position, external_id, title, artist, album, duration_s,
            candidates, status)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, '[]', 'pending')`,
        [
          randomHex(ID_BYTES),
          importId,
          position,
          track.external_id ?? null,
          track.title,
          track.artist,
          track.album ?? null,
          track.duration_s ?? null,
        ],
      )
    }
    await db.runAsync(
      `UPDATE playlist_imports
         SET track_count = (SELECT COUNT(*) FROM track_matches WHERE playlist_import_id = ?),
             updated_at = ?
       WHERE id = ?`,
      [importId, timestamp, importId],
    )
  })
  return tracks.length
}

export async function listMatches(
  importId: string,
  status?: LocalTrackMatchStatus,
): Promise<LocalTrackMatch[]> {
  const db = await openLibraryDb()
  const rows = status
    ? await db.getAllAsync<TrackMatchRow>(
        `SELECT * FROM track_matches WHERE playlist_import_id = ? AND status = ?
         ORDER BY position`,
        [importId, status],
      )
    : await db.getAllAsync<TrackMatchRow>(
        `SELECT * FROM track_matches WHERE playlist_import_id = ? ORDER BY position`,
        [importId],
      )
  return rows.map(hydrate)
}

export async function getMatch(matchId: string): Promise<LocalTrackMatch | null> {
  const db = await openLibraryDb()
  const row = await db.getFirstAsync<TrackMatchRow>(`SELECT * FROM track_matches WHERE id = ?`, [
    matchId,
  ])
  return row ? hydrate(row) : null
}

/** How many matches an import has in each status, for the progress counters. */
export async function countMatchesByStatus(importId: string): Promise<Record<string, number>> {
  const db = await openLibraryDb()
  const rows = await db.getAllAsync<{ status: string; total: number }>(
    `SELECT status, COUNT(*) AS total FROM track_matches
     WHERE playlist_import_id = ? GROUP BY status`,
    [importId],
  )
  return Object.fromEntries(rows.map((row) => [row.status, row.total]))
}

/** Record what a search found for one track, and what the scorer made of it. */
export async function setMatchCandidates(
  matchId: string,
  candidates: readonly LocalMatchCandidate[],
  status: LocalTrackMatchStatus,
): Promise<void> {
  const db = await openLibraryDb()
  const best = candidates.length > 0 ? candidates[0] : null
  await db.runAsync(
    `UPDATE track_matches SET candidates = ?, chosen_url = ?, confidence = ?, status = ?
     WHERE id = ?`,
    [
      JSON.stringify(candidates),
      // Only a match good enough to stand on its own is pre-chosen; the rest
      // are offered to the human, which is what `review` means.
      status === 'auto_matched' && best ? best.url : null,
      status === 'auto_matched' && best ? best.score : null,
      status,
      matchId,
    ],
  )
}

/**
 * The human's answer for one track.
 *
 * ⚠️ **`confidence` is derived here, not passed in.** It means *machine*
 * confidence, so it can only ever be the score of a candidate the scorer
 * actually produced: this looks the chosen URL up among the stored candidates
 * and stores that candidate's score, or `null` when the URL is not one of them
 * — which is exactly the hand-pasted case.
 *
 * Taking it as an argument, as the first version of this did, made the rule the
 * *caller's* to keep and left nothing able to fail when it stopped keeping it.
 * A guarantee beats a prohibition.
 */
export async function chooseMatch(matchId: string, chosenUrl: string | null): Promise<void> {
  const db = await openLibraryDb()
  if (chosenUrl === null) {
    await db.runAsync(
      `UPDATE track_matches SET chosen_url = NULL, confidence = NULL, status = 'rejected'
       WHERE id = ?`,
      [matchId],
    )
    return
  }
  const existing = await getMatch(matchId)
  const chosen = existing?.candidates.find((candidate) => candidate.url === chosenUrl)
  await db.runAsync(
    `UPDATE track_matches SET chosen_url = ?, confidence = ?, status = 'accepted' WHERE id = ?`,
    [chosenUrl, chosen?.score ?? null, matchId],
  )
}

export async function setMatchStatus(
  matchId: string,
  status: LocalTrackMatchStatus,
  patch: { song_id?: string | null; error?: string | null } = {},
): Promise<void> {
  const db = await openLibraryDb()
  await db.runAsync(`UPDATE track_matches SET status = ?, song_id = ?, error = ? WHERE id = ?`, [
    status,
    patch.song_id ?? null,
    patch.error ?? null,
    matchId,
  ])
}

/**
 * Delete an import and its matches.
 *
 * No `ON DELETE CASCADE`: SQLite honours foreign keys only when
 * `PRAGMA foreign_keys` is on, which is off by default and per-connection, so a
 * declared cascade that silently does nothing is worse than deleting by hand.
 * `playlists.ts` reached the same conclusion for the same reason.
 */
export async function deleteImport(importId: string): Promise<void> {
  const db = await openLibraryDb()
  await withLibraryTransaction(db, async () => {
    await db.runAsync(`DELETE FROM track_matches WHERE playlist_import_id = ?`, [importId])
    await db.runAsync(`DELETE FROM playlist_imports WHERE id = ?`, [importId])
  })
}

/** One page of matches, shaped like the server's `Page<T>` so the review
 *  screen's infinite query does not have to change (#611). */
export interface MatchPage {
  items: LocalTrackMatch[]
  total: number
  limit: number
  offset: number
}

/**
 * A page of matches, optionally filtered.
 *
 * SQLite could return all of them at once and the screen would work — but the
 * review list is an infinite query over a playlist that can be two hundred
 * tracks, and `MatchReview.tsx` is built around paging. Keeping the shape keeps
 * that file untouched, which is the whole test of whether this port drifted.
 */
export async function pageMatches(
  importId: string,
  status: LocalTrackMatchStatus | 'all',
  limit: number,
  offset: number,
): Promise<MatchPage> {
  const db = await openLibraryDb()
  const filtered = status !== 'all'
  const where = `WHERE playlist_import_id = ?${filtered ? ' AND status = ?' : ''}`
  const params = filtered ? [importId, status] : [importId]

  const counted = await db.getFirstAsync<{ total: number }>(
    `SELECT COUNT(*) AS total FROM track_matches ${where}`,
    params,
  )
  const rows = await db.getAllAsync<TrackMatchRow>(
    `SELECT * FROM track_matches ${where} ORDER BY position LIMIT ? OFFSET ?`,
    [...params, limit, offset],
  )
  return { items: rows.map(hydrate), total: counted?.total ?? 0, limit, offset }
}

/** How many tracks confirming would actually download: the matcher was
 *  confident and the user left it alone, or the user said yes. */
export async function importableCount(importId: string): Promise<number> {
  const counts = await countMatchesByStatus(importId)
  return (counts.auto_matched ?? 0) + (counts.accepted ?? 0)
}

/** Set the same status on several matches at once — the only way a hundred-track
 *  playlist is reviewable on a phone at all. */
export async function setMatchStatuses(
  matchIds: readonly string[],
  status: LocalTrackMatchStatus,
): Promise<void> {
  if (matchIds.length === 0) return
  const db = await openLibraryDb()
  await withLibraryTransaction(db, async () => {
    for (const matchId of matchIds) {
      await db.runAsync(`UPDATE track_matches SET status = ? WHERE id = ?`, [status, matchId])
    }
  })
}

/**
 * Put failed tracks back in the queue (#223).
 *
 * Only the ones recorded as failed — the successful tracks keep their songs, so
 * this is not a second import of the whole playlist. Their error is cleared
 * with them, or the row would carry a reason for a failure that is no longer
 * true.
 */
export async function retryFailedMatches(importId: string): Promise<number> {
  const db = await openLibraryDb()
  const failed = await db.getAllAsync<{ id: string }>(
    `SELECT id FROM track_matches WHERE playlist_import_id = ? AND status = 'failed'`,
    [importId],
  )
  await db.runAsync(
    `UPDATE track_matches SET status = 'accepted', error = NULL
     WHERE playlist_import_id = ? AND status = 'failed'`,
    [importId],
  )
  await updateImport(importId, { status: 'importing', failed_count: 0 })
  return failed.length
}

/**
 * The import that already exists for this playlist, if there is one.
 *
 * What makes re-importing a living playlist fill in the one already here rather
 * than build a second beside it — the same argument the `google_playlist_id`,
 * `bilibili_fav_id` and `bilibili_video_id` columns make for playlists (v7–v9).
 */
export async function findImport(
  service: ImportService,
  externalPlaylistId: string,
): Promise<LocalPlaylistImport | null> {
  const db = await openLibraryDb()
  return db.getFirstAsync<LocalPlaylistImport>(
    `SELECT * FROM playlist_imports WHERE service = ? AND external_playlist_id = ?
     ORDER BY created_at DESC LIMIT 1`,
    [service, externalPlaylistId],
  )
}

/**
 * The local playlist this import fills, creating it on first use.
 *
 * Replaces `playlistForServerImport`, which keyed a playlist to the **server's
 * integer** import id via `playlists.server_import_id` — a column local imports
 * cannot use, because their ids are TEXT (#610).
 *
 * ⚠️ **This needs no new column on `playlists`.** v10 already carries
 * `playlist_imports.playlist_id`, so the link exists in the other direction,
 * and re-importing a living playlist reuses the row that `findImport` turns up
 * rather than building a second one beside it — the same argument v7–v9 make
 * with `google_playlist_id`, `bilibili_fav_id` and `bilibili_video_id`.
 */
export async function playlistForImport(importId: string, name: string): Promise<string> {
  const existing = await getImport(importId)
  if (existing?.playlist_id) {
    const db = await openLibraryDb()
    const playlist = await db.getFirstAsync<{ id: string }>(
      `SELECT id FROM playlists WHERE id = ?`,
      [existing.playlist_id],
    )
    // A playlist the user deleted by hand leaves a dangling id. Making a fresh
    // one is the only sane answer; refusing to import would punish them for
    // tidying up.
    if (playlist) return playlist.id
  }
  const playlistId = await createPlaylist(name)
  await updateImport(importId, { playlist_id: playlistId })
  return playlistId
}
